### PR TITLE
[6X backport] Fix the double-free issue of work_set

### PR DESCRIPTION
**Pull request by gfphoenix78**
The workfile_set(or work_set referenced in the code) is used to
track the usage of the temporary files. For example, in hash join,
if the hash table is too large to put them all in the memory,
the hash table is split and saved in temporary files. When the
transaction is finished(no matter commit or abort), the workfile_set
referenced by the executor node will be freed(reference count).
There are also other workfile_set with different lifecycle(e.g. session).
The backend records the usage of the workfile_set in its local memory,
and it will free all the `workfile_set`s in the callback(AtProcExit_WorkFile)
when the backend is going to exit.

The problem is that the function `AtProcExit_WorkFile` is called before
the cleanup of the current transaction. So, the executor node may free
the workfile_set after all `workfile_set` in the current backend have been
freed.
The following backtrace shows a PANIC issue:
```
2021-05-12 01:35:59.902395 UTC,"gpadmin","gpadmin",p11681,th-22202240,"127.0.0.1","15174",2021-05-12 01:35:56 UTC,0,con10,cmd4,seg-1,,dx5,,sx1,"PANIC","XX000","workfile_set is not active",,,,,,,0,,"workfile_mgr.c",697,"Stack trace:
1    0xe0c7e4 postgres errstart (elog.c:557)
2    0xe99372 postgres workfile_mgr_close_set (workfile_mgr.c:697)
3    0xa8f36c postgres ExecHashTableDestroy (nodeHash.c:694)
4    0xa9347a postgres ExecEndHashJoin (nodeHashjoin.c:763)
5    0xa6b335 postgres ExecEndNode (execProcnode.c:1444)
6    0xa934ee postgres ExecEndHashJoin + 0xcc
7    0xa6b335 postgres ExecEndNode + 0x2ff
8    0xa8e7c5 postgres ExecEndHash + 0x30
9    0xa6b3ac postgres ExecEndNode + 0x376
10   0xa934fe postgres ExecEndHashJoin + 0xdc
11   0xa6b335 postgres ExecEndNode + 0x2ff
12   0xa64d7e postgres <symbol not found> (discriminator 2)
13   0xa61e3e postgres standard_ExecutorEnd + 0x381
14   0xa61aba postgres ExecutorEnd + 0x36
15   0x9ef89a postgres PortalCleanup (portalcmds.c:326)
16   0xe52e54 postgres AtAbort_Portals (portalmem.c:837)
17   0x8626ac postgres <symbol not found> (xact.c:3297)
18   0x864e02 postgres AbortOutOfAnyTransaction + 0x6e
19   0xe2894d postgres <symbol not found> (postinit.c:1348)
20   0xc6c672 postgres shmem_exit (ipc.c:256)
21   0xc6c567 postgres proc_exit_prepare + 0xcb
22   0xc6c465 postgres proc_exit (ipc.c:143)
23   0xca35ea postgres PostgresMain (postgres.c:5679)
24   0xc0f41c postgres <symbol not found> (postmaster.c:5442)
25   0xc0eaad postgres <symbol not found> + 0xc0eaad
26   0xc0a8ef postgres <symbol not found> + 0xc0a8ef
27   0xc09e3b postgres PostmasterMain + 0x12e5
28   0xb02495 postgres <symbol not found> (main.c:264)
29   0x7fb5fbb94505 libc.so.6 __libc_start_main + 0xf5
30   0x7b4729 postgres <symbol not found> + 0x7b4729
"
```

The issue occurs in JDBC. The driver uses extended query, after running
`exec_bind_message()` and `exec_execute_message()`, the portal is still
alive and its status is PORTAL_READY. When the user presses Ctl-C to
kill the session, the backend starts to do some cleanup and exit.

This commit reorder the cleanup function of the workfile_set(`AtProcExit_WorkFile`).
So, the `workfile_set` referenced in the executor node will be freed first,
then `AtProcExit_WorkFile` frees all remaining `workfile_set`s used by
the current backend.

Co-authored-by: Gang Xiong <gangx@vmware.com>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by gfphoenix78** [comment]
The steps to reproduce this PANIC are:

The steps are:

* Prepare dataset:
```
1) create a user table
create table foo1( i int); -- used by java program
2) create a lot of schema in a database, in our environment, there are about 100,000 schemas.
```
* Prepare a JDBC program to run the query: SELECT EXISTS ( SELECT 1 FROM information_schema.tables WHERE table_schema = 'public' AND table_name ='foo1' )
```
1) Download the JDBC driver at https://jdbc.postgresql.org/download/postgresql-42.2.20.jar
2) Write a JDBC program and a helper script file, see the attachment files.
```
* Set the GUC statement_mem=1024
* Run the java program bash run.bash and wait until it outputs start to sleep
* Press Ctl-C to kill the above command(step 4).

**Comment by gfphoenix78** [comment]
Main.java
```java
import java.sql.Connection;
import java.sql.DatabaseMetaData;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Enumeration;

/**
 * Created by wangbin on 2017/8/19.
 */

public class Main {
    public static void main(String[] args) throws ClassNotFoundException, SQLException, Exception {
        Connection con;

        // shop 为数据库名
        //String url = "jdbc:postgresql://localhost:6000/gpadmin";
        String url = "jdbc:postgresql://localhost:5432/gpadmin";

        // user 为数据库 shop 的拥有者名
        String user = "gpadmin";
        String password = "";

        // 加载 JDBC 驱动器
        Class.forName("org.postgresql.Driver");

        Enumeration e = DriverManager.getDrivers();

        for(;e.hasMoreElements();) {
            System.out.println(e.nextElement().getClass());
        }

        // 建立连接
        con = DriverManager.getConnection(url, user, password);


        DatabaseMetaData dbmt = con.getMetaData();
        System.out.println(dbmt.getURL() + "\t" + dbmt.getDriverName() + "\t"
                + dbmt.getDriverVersion() + "\t"
                + dbmt.getDatabaseProductName() + "\t"
                + dbmt.getDatabaseProductVersion());

        test3(con);

        con.close();
    }
    public static void test3(Connection conn) throws SQLException, InterruptedException {
        conn.setAutoCommit(false);
        String sql = "SELECT EXISTS ( SELECT 1 FROM information_schema.tables WHERE table_schema = 'public' AND table_name = 'foo1' )";
        Statement st = conn.createStatement();
        ResultSet rs = st.executeQuery(sql);
        int num = 0;
        while (rs.next()) {
            num++;
        }
        System.out.println("result: => " + num);
        System.out.println("start to sleep");
        Thread.sleep(36000 *1000);
        rs.close();
        st.close();
    }
}
```

run.bash
```bash
#!/bin/bash
javac Main.java && java -cp .:postgresql-42.2.20.jar Main
```

**Review by asimrp — COMMENTED** [review]
Thank you for the detailed explanation in commit message as well as the JDBC example.  The change looks good :+1:

How can we add a test to validate this?  The same combination of `exec_bind_message()` and `exec_execute_message()` should be achieved by using libpq `PQprepare()` followed by `PQexecPrepared()` interface.  We already have an example to refer to in isolation2 tests - `isolation2/extended_protocol_test.c`.  Do you think it's worth trying?

We should first fix this in Greenplum 7 and then back port to 6X, unless there is a reason not to fix this in 7.

**Comment by asimrp** [comment]
Ah, just realized that you have a PR on Greenplum 7 too: #11965.

**Comment by gfphoenix78** [comment]
> Thank you for the detailed explanation in commit message as well as the JDBC example. The change looks good 👍
> 
> How can we add a test to validate this? The same combination of `exec_bind_message()` and `exec_execute_message()` should be achieved by using libpq `PQprepare()` followed by `PQexecPrepared()` interface. We already have an example to refer to in isolation2 tests - `isolation2/extended_protocol_test.c`. Do you think it's worth trying?
> 
> We should first fix this in Greenplum 7 and then back port to 6X, unless there is a reason not to fix this in 7.

I'd like to add the test case, but the hash join node doesn't allocate the workfile_set when I run the external c program suggested by Zhenghua. I'll appreciate it if someone could help me with the test case.

**Review by paul-guo- — APPROVED** [review]
Probably deserves a try to add a test case. e.g. use some planner guc to enforce a hash join with small table size + stat memory setting + extended query program.

File: src/backend/utils/workfile_manager/workfile_mgr.c [hunk 1]
```diff
@@ -547,7 +547,7 @@ workfile_mgr_create_set(const char *operator_name, const char *prefix, bool hold
 	if (!proc_exit_hook_registered)
 	{
 		/* register proc-exit hook to ensure temp files are dropped at exit */
-		before_shmem_exit(AtProcExit_WorkFile, 0);
```
Comment (paul-guo-) [inline src/backend/utils/workfile_manager/workfile_mgr.c:550]: Can we comment a bit here? To avoid some people changing it back to `before_shmem_exit()`?

**Comment by gfphoenix78** [comment]
Why I missed the test case?
Well, I intended to add a test case, but found it hard.
First, the repro steps require running the extended query(prepare/execute), so the portal keeps alive until the end of the session and the portal's status is not active when the error happens.
Second, the inner table of the hash join must be relatively big to spill the hash table to the disk, so workfile_set is allocated.

The initial repro steps create tens of thousands of schemas/tables which makes the pg_namespace and pg_type big enough to spill with small statement_mem. I'm able to repro this issue steadily. However, creating so many schemas/tables in a single test seems tricky and hurts the pipeline time.

Next, I try to join 2 user tables:
```
create table foo1(i int);
create table foo2(i int, j int);
-- insert 10K tuples to foo1 and foo2
gpadmin=# explain select f1.i, f2.j from foo1 f1, foo2 f2 where f1.i=f2.i;
                                    QUERY PLAN
-----------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..916.88 rows=100000 width=8)
   ->  Hash Join  (cost=0.00..907.94 rows=100000 width=8)
         Hash Cond: (foo2.i = foo1.i)
         ->  Seq Scan on foo2  (cost=0.00..433.09 rows=100000 width=8)
         ->  Hash  (cost=432.87..432.87 rows=100000 width=4)
               ->  Seq Scan on foo1  (cost=0.00..432.87 rows=100000 width=4)
 Optimizer: Pivotal Optimizer (GPORCA)
(7 rows)
select f1.i, f2.j from foo1 f1, foo2 f2 where f1.i=f2.i;
```
The external C program runs the above SELECT query, but I still couldn't repro this bug. Because the portal in the QE process is freed after the current command before the end of the transaction. With no active portal in the QE, the bug can't occur.

**Pull request merged**